pkg/recovery/etcd.go

```diff
@@ -0,0 +1,147 @@
+// The etcd backend fetches control plane objects directly from etcd. This is adapted heavily from
+// kubernetes/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go.
+
+package recovery
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/pkg/api"
+)
+
+// EtcdBackend is a backend that extracts a controlPlane from an etcd instance.
+type EtcdBackend struct {
+	client     *clientv3.Client
+	decoder    runtime.Decoder
+	pathPrefix string
+}
+
+// NewEtcdBackend constructs a new EtcdBackend for the given client and pathPrefix.
+func NewEtcdBackend(client *clientv3.Client, pathPrefix string) Backend {
+	return &EtcdBackend{
+		client:     client,
+		decoder:    api.Codecs.UniversalDecoder(),
+		pathPrefix: pathPrefix,
+	}
+}
+
+// read implements Backend.read().
+func (s *EtcdBackend) read(ctx context.Context) (*controlPlane, error) {
+	cp := &controlPlane{}
+	for _, r := range []struct {
+		etcdName string
+		yamlName string
+		obj      runtime.Object
+	}{{
+		etcdName: "configmaps",
+		yamlName: "config-map",
+		obj:      &cp.configMaps,
+	}, {
+		etcdName: "daemonsets",
+		yamlName: "daemonset",
+		obj:      &cp.daemonSets,
+	}, {
+		etcdName: "deployments",
+		yamlName: "deployment",
+		obj:      &cp.deployments,
+	}, {
+		etcdName: "poddisruptionbudgets",
+		yamlName: "pod-disruption-budget",
+		obj:      &cp.podDisruptionBudgets,
+	}, {
+		etcdName: "secrets",
+		yamlName: "secret",
+		obj:      &cp.secrets,
+	}, {
+		etcdName: "services/specs",
+		yamlName: "service",
+		obj:      &cp.services,
+	}, {
+		etcdName: "serviceaccounts",
+		yamlName: "service-account",
+		obj:      &cp.serviceAccounts,
+	},
```

Comment:
I guess my leaning toward being more specific about what we're recovering is that this wouldn't technically cover everything in a cluster - as we're only recovering some of the assets (e.g. there could scheduled jobs, stateful sets, namespaces, rbac rules, others that I'm not thinking of).

If we were really trying to recover all cluster assets / stamp out a new cluster (now that I'm revisiting this idea) we would need to get all the objects that *could* exist. Even then - we would need to introduce ordering to object creation (e.g. namespaces need to be created prior to objects that exist in those namespaces, or TPR types before instances of those types).

Regarding, recovering from etcd-backup - my assumption is that we would only support this for self-hosted etcd(?) - in which case the data would already be replicated into the cluster (essentially by re-executing the pivot into a new cluster), so we wouldn't necessarily need to re-create the assets (they're still there).

Ultimately, dynamically extracting all object types - then enforcing ordering constraints isn't insurmountable -- but it does seem like a decent amount of complexity vs starting out with a more focused set of assets to recover.